### ISSUE TITLE
Need help for dynamic pass params 

### ISSUE THREAD

**Issue by ganatrajay** [B3nedikt/restring#25]
hi,

i have response in json like language code as key map<"String","String"> obj need to pass on runtime in Resting and how to fetch and set for textview. please pass me one example 

Regards,
jay

**Comment by B3nedikt** [comment]
I need more details to answer your question:

- Do you need a kotlin or java example?
- Do you only want to set the the new strings for a single TextView? While this is possible, the default behavior of restring is the set new strings for all types of views.

If you have some code what you are trying to achieve, this would be useful. You can just fork this repository and modify the sample app to what you think you need if you want. Then I can point you in the right direction easier ;)

**Comment by ganatrajay** [comment]
[DynamicLayoutApp.zip](https://github.com/B3nedikt/restring/files/4675101/DynamicLayoutApp.zip)

**Comment by B3nedikt** [comment]
As a generel note, don´t upload zip files of source codes, create a repository & link it. This is kind of a security issue. Also when you send demo code, format it properly, and remove uncommented code. Literally half of your MainActivity consists of uncommented code.

That aside, your BaseActivity is wrong, just copy the code from this repositories Readme or the sample app. Also you are doing "applicationContext.resources.getString(...)". This should be just "getString(...)", as you need to context you wrapped in the BaseActivity, not the applicationContext. Then your code should work. Hope this helps!

**Comment by ganatrajay** [comment]
Hi,

Sorry for above link, ya i was using https://github.com/hamidness/restring before but was not working in android 7 API 24 nougat version so code was messed up.   here i have attached the files as you said in demo app similarly i have applied changes but still not working. i hope we just need to pass like textview.text = getString(R.string.title) and this title will be in string.xml like <string name="title"></string> (empty key)


[demoApp.txt](https://github.com/B3nedikt/restring/files/4676146/demoApp.txt)

**Comment by B3nedikt** [comment]
No problem, sharing your code with github is just way easier. If you don´t know how there are multiple tutorials right here on github!
There are two problems with your code:

- You use both the StringsLoader mechanism, as well as the Restring.setStrings(...) method. While this works, these two override each other, so you need to decide on one! I would recommend to just move the chinese support code to the StringsLoader.
- You are using my AppLocale lib also. This lib overwrites the default method how the app language is determined.
So this line of code:
`
AppLocale.supportedLocales = listOf(Locale.ENGLISH, Locale.US, LOCALE_AUSTRIAN_GERMAN)
`
does constraint the app to only these languages. So you need to add chinese to it: 
`
AppLocale.supportedLocales = listOf(Locale.ENGLISH, Locale.US, LOCALE_AUSTRIAN_GERMAN, Locale.Chinese)
`

**Comment by ganatrajay** [comment]
Hi,

i got your point but i am getting labels from api service, so on runtime i have to add language Locale (getting language iso code from service api) and have to add key,value map in Restring.setStrings(Locale,mapObject). 

- first can i add this line AppLocale.supportedLocales = listOf(Locale.ENGLISH, Locale.US, LOCALE_AUSTRIAN_GERMAN) on runtime not in application level ?

- second Restring.setString(locale, "key","value") works fine but Restring.setStrings(locale,mapObject) is not working . 

please do let me know if i can share demo sample app, anywhere . thanks

**Comment by B3nedikt** [comment]
You can also call an API in the StringsLoader, so you can get the language code right there if you want!

About the other questions:
- first can i add this line AppLocale.supportedLocales = listOf(Locale.ENGLISH, Locale.US, LOCALE_AUSTRIAN_GERMAN) on runtime not in application level ?

==> Yes

- second Restring.setString(locale, "key","value") works fine but Restring.setStrings(locale,mapObject) is not working .

==> Restrings overwrite behavior is the problem here I think. Restring.setStrings(..) overwrites whatever you set for this language, which is not really a good behavior. I may change this in a later release when I find the time.

The best way to share the sample app, would be as a git repository. Just upload your code right here on github!

**Comment by ganatrajay** [comment]
Hi,

here i have attached code snippet again, facing reflect issue. 

- i have given my api response as well below , i will wait for you next release to work on Restring.setStrings because i have to add key as language and value as map of many lables(key,value)

- another issue is from application file only i have to add supported locales, please give me another way where i can make. 

- after adding Restring.setString(locale,"key","value") runtime ,  from spinner if we change the language code, its not reflecting immediately we have to change orientation 

please let me know how i can do changes thanks 
[demoApp.txt](https://github.com/B3nedikt/restring/files/4676781/demoApp.txt)

**Comment by B3nedikt** [comment]
I fixed the overwrite issue in the latest version of the lib. Just update the lib to 4.0.2 and everything should work. I would recommend using setStrings(..) though as this will be more efficient then adding every string one after an other.

So your MainActivity should look like this:
```kotlin
class MainActivity : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        
        AppLocale.supportedLocales = listOf(Locale.ENGLISH, Locale.US, Locales.LOCALE_AUSTRIAN_GERMAN, Locale.CHINESE)

        val map: Map<String, String> = mapOf(
                "title" to "Title Chinese (ZH restring)",
                "subtitle" to "Subtitle Chinese (ZH restring)",
                "text_view_text" to "TextView Text Chinese"
        )

        Restring.setStrings(Locale.CHINESE, map)

        val localeStrings = AppLocale.supportedLocales.map { it.language + " " + it.country }
        val adapter = ArrayAdapter(this,
                android.R.layout.simple_dropdown_item_1line, localeStrings)

        spinner.adapter = adapter
    }

    override fun onResume() {
        super.onResume()

        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                AppLocale.desiredLocale = AppLocale.supportedLocales[position]

                val rootView = window.decorView.findViewById<ContentFrameLayout>(android.R.id.content)
                Reword.reword(rootView)

                stringArrayTextView.text = resources.getStringArray(R.array.string_array)
                        .joinToString("\n")

                quantityStringTextView.text = (0 until 3)
                        .joinToString("\n")
                        { resources.getQuantityString(R.plurals.quantity_string, it, it) }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) = Unit
        }
    }
}
```
Hope this helps :)

**Comment by ganatrajay** [comment]
Hi,

Thank you so much for quick reply and updated version, everything works fine now except new added map object dynamically is not appearing at first time , i have to rotate the device then it comes up . i m going to attach same given sample code please try it out thanks. 


[restring-master.zip](https://github.com/B3nedikt/restring/files/4686712/restring-master.zip)

**Comment by B3nedikt** [comment]
Your´re welcome :)
For me your sample code seems to work, when I switch the spinner to "zh" I get the strings for Chinese and not the ones for English as expected. Does the issue appear to you independent of the phone you are using? Maybe it is a phone specific issue.

**Comment by B3nedikt** [comment]
For clarity, these to lines:
````
val rootView = window.decorView.findViewById<ContentFrameLayout>(android.R.id.content)
Reword.reword(rootView)
````
you see in the code of the spinner trigger the language change!

**Comment by ganatrajay** [comment]
Hi, 

thanks for code snippet but i tried to run this in multiple devices like android 9 , 10 even in emulator version Api 24 Android Nougat version also. facing the same issue. i am just stuck on this point, please help to resolve this. 

here i have attached the screenshots for more information on issue part where you can see on idle portrait mode if you select chinese as its added on runtime in setStrings() its not reflecting and have to turn device in other orientation so it get recreate the activity or configuration change calls so it appears and then its working fine. 

![portrait_mode](https://user-images.githubusercontent.com/18435452/83100835-77182680-a0ce-11ea-86f7-ef240fc46aeb.PNG)
![landscape](https://user-images.githubusercontent.com/18435452/83100842-78e1ea00-a0ce-11ea-84a5-cb5a97956739.PNG)

Will wait for your reply thanks

**Comment by B3nedikt** [comment]
I think I found the issue, and will probably implement a fix later to day.

Just to be sure: Please remove the the line with AppLocale.supportedLocales = listOf(...) & the line with .stringsLoader(...) from the Application class.
It should then look like this:
````kotlin
class SampleApplication : Application() {

    override fun onCreate() {
        super.onCreate()

        Restring.init(this,
                RestringConfig.Builder()
                        .localeProvider(AppLocaleLocaleProvider)
                        .build()
        )

        ViewPump.init(ViewPump.builder()
                .addInterceptor(RewordInterceptor)
                .build()
        )
    }
}
```
But leave it there in the MainActivity!
This should lead to only Chinese strings getting replaced, and all other languages not. If this works for you like this, the issue is as expected, and my next fix should make it work for you :)

**Comment by B3nedikt** [comment]
I have released version 4.0.3, I think this should now finally fix the issue ;)

**Comment by ganatrajay** [comment]
Hey thank you so much for great help , it works cool :)
